sql/src/test/java/io/crate/execution/engine/aggregation/GroupBySingleNumberCollectorTest.java

```diff
@@ -0,0 +1,94 @@
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.engine.aggregation;
+
+import io.crate.breaker.RamAccountingContext;
+import io.crate.data.BatchIterator;
+import io.crate.data.BatchIterators;
+import io.crate.data.InMemoryBatchIterator;
+import io.crate.data.Input;
+import io.crate.data.Row;
+import io.crate.data.Row1;
+import io.crate.execution.engine.aggregation.impl.AggregationImplModule;
+import io.crate.execution.engine.aggregation.impl.SumAggregation;
+import io.crate.execution.engine.collect.CollectExpression;
+import io.crate.execution.engine.collect.InputCollectExpression;
+import io.crate.expression.symbol.AggregateMode;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.Functions;
+import io.crate.test.integration.CrateUnitTest;
+import io.crate.types.DataTypes;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.common.breaker.NoopCircuitBreaker;
+import org.elasticsearch.common.inject.ModulesBuilder;
+import org.elasticsearch.common.util.BigArrays;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static io.crate.data.SentinelRow.SENTINEL;
+import static org.hamcrest.Matchers.is;
+
+public class GroupBySingleNumberCollectorTest extends CrateUnitTest {
+
+    private static final RamAccountingContext RAM_ACCOUNTING_CONTEXT =
+        new RamAccountingContext("dummy", new NoopCircuitBreaker(CircuitBreaker.FIELDDATA));
+
+    @Test
+    public void testIterateOverResultTwice() throws Exception {
+        Functions functions = new ModulesBuilder().add(new AggregationImplModule())
+            .createInjector().getInstance(Functions.class);
+        AggregationFunction sumAgg = (AggregationFunction) functions.getQualified(
+            new FunctionIdent(SumAggregation.NAME, Arrays.asList(DataTypes.INTEGER)));
+        InputCollectExpression keyInput = new InputCollectExpression(0);
+
+        GroupBySingleNumberCollector groupBySingleNumberCollector = new GroupBySingleNumberCollector(
+            DataTypes.INTEGER,
+            new CollectExpression[]{keyInput},
+            AggregateMode.ITER_FINAL,
+            new AggregationFunction[]{sumAgg},
+            new Input[][]{new Input[]{keyInput}},
+            RAM_ACCOUNTING_CONTEXT,
+            keyInput,
+            Version.CURRENT,
+            BigArrays.NON_RECYCLING_INSTANCE
+        );
+
+        BatchIterator<Row> inputRowsIterator = InMemoryBatchIterator.of(
+            IntStream.range(0, 10).mapToObj(Row1::new).collect(Collectors.toList()),
+            SENTINEL
+        );
+
+        Iterable<Row> rows = BatchIterators.collect(inputRowsIterator, groupBySingleNumberCollector).get();
```

Comment:
Maybe use the `get` overload with a timeout, so that if we break something, the test doesn't get stuck but fails eventually. (If you run all tests and one gets stuck, it is tricky so see which one).